### PR TITLE
python3Packages.py-air-control: init at 2.1.0

### PR DESCRIPTION
**Pull request by urbas**
###### Motivation for this change

This Python package was not yet packaged in nixpkgs.

Package size:
```
$ nix path-info -sh $(nix-build . -A python3Packages.py-air-control)
/nix/store/07ssfnqs4gp89pscij442nd51bvdj771-python3.8-py-air-control-2.1.0       116.9K
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @SuperSandro2000

### CODE REVIEW

**Comment by urbas** [comment]
cc @SuperSandro2000

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please disable it on python27 and disable the tests because the package has none.

```
running egg_info
writing py_air_control.egg-info/PKG-INFO
writing dependency_links to py_air_control.egg-info/dependency_links.txt
writing entry points to py_air_control.egg-info/entry_points.txt
writing requirements to py_air_control.egg-info/requires.txt
writing top-level names to py_air_control.egg-info/top_level.txt
reading manifest file 'py_air_control.egg-info/SOURCES.txt'
writing manifest file 'py_air_control.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
Finished executing setuptoolsCheckPhase
```

**Comment by urbas** [comment]
@SuperSandro2000: disabled the tests and on python27.

I also took the liberty and did the same for the `coapthon3` package (which I added recently). I did so in separate commits though.

File: pkgs/development/python-modules/coapthon3/default.nix [hunk 1]
```diff
@@ -11,6 +11,9 @@ buildPythonPackage rec {
 
   propagatedBuildInputs = [ cachetools ];
 
+  # Disabling tests because they are not included in the package.
+  doCheck = false;
+
```
Comment (lukegb) [inline pkgs/development/python-modules/coapthon3/default.nix:16]: Having disabled tests, consider adding

```nix
  pythonImportsCheck = [ "coapthon" ];
```

which will at least validate that it's possible to import the package.

File: pkgs/development/python-modules/py-air-control/default.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0g9qkaygjyyxkhgx51m8hhj9inlzmynzw6l4vp7g1cflrn8g3i92";
+  };
+
+  propagatedBuildInputs = [ pycryptodomex coapthon3 ];
+
+  # Disabling tests because they are not included in the package.
+  doCheck = false;
```
Comment (lukegb) [inline pkgs/development/python-modules/py-air-control/default.nix:15]: Ditto here (please add a `pythonImportsCheck`)

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -1255,7 +1255,7 @@ in {
 
   cnvkit = callPackage ../development/python-modules/cnvkit { };
 
-  coapthon3 = callPackage ../development/python-modules/coapthon3 { };
+  coapthon3 = disabledIf isPy27 (callPackage ../development/python-modules/coapthon3 { });
```
Comment (lukegb) [inline pkgs/top-level/python-packages.nix:1258]: Please do this in the package itself, e.g.:

```nix
  disabled = isPy27;
```

conventionally directly under the `version =`

File: pkgs/development/python-modules/py-air-control/default.nix [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{ buildPythonPackage, coapthon3, fetchPypi, lib, pycryptodomex }:
```
Comment (lukegb) [inline pkgs/development/python-modules/py-air-control/default.nix:1]: Is this useful as a Python module, or is it an application (i.e. is it really just for the "airctrl" CLI)?

If the former, this is great :)

If the latter, consider putting this elsewhere in the tree and then using "python3Packages.callPackage" from "all-packages.nix" instead.

**Review by lukegb — CHANGES_REQUESTED** [review]
A few things to ask you...

File: pkgs/development/python-modules/py-air-control/default.nix (hunk 4 above)
Comment (urbas) [reply pkgs/development/python-modules/py-air-control/default.nix:1]: It's the former.

To give more context: I am using `py-air-control` as a library in [another Python module](https://github.com/urbas/py-air-control-exporter). Here's [an example of how it's used](https://github.com/urbas/py-air-control-exporter/blob/master/py_air_control_exporter/metrics.py#L5).

The final goal is to continuously measure air quality, publish the measurements to Prometheus, and visualize in Grafana (all in NixOS on a Raspberry Pi). I hope this makes sense.

Comment (lukegb) [reply pkgs/development/python-modules/py-air-control/default.nix:1]: Sure! Nice :)

(ooh, I really want to start doing air quality measurement...)

File: pkgs/development/python-modules/py-air-control/default.nix [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0g9qkaygjyyxkhgx51m8hhj9inlzmynzw6l4vp7g1cflrn8g3i92";
+  };
+
+  propagatedBuildInputs = [ pycryptodomex coapthon3 ];
+
+  # Disabling tests because they are not included in the package.
+  doCheck = false;
```
Comment (jonringer) [inline pkgs/development/python-modules/py-air-control/default.nix:16]: Please see if the github repo has tests

There are no tests included. Please try to checkout from source and check if they have unit tests, and try to run them. Unit tests give a good indication that they package has a high degree of validity and correctness given the python package set.

If tests are not available, then please use `pythonImportsCheck` to import the most important modules. This isn't as good as unit tests, but will usually give a good indication of run-time errors. Please see [pythonImportsCheck documentation](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonimportscheck) for more information.

Comment (urbas) [reply pkgs/development/python-modules/py-air-control/default.nix:16]: Oh, I was wondering about that. Yes, there are tests in Github (just not distributed in PyPi). I thought that we should always use `fetchPypi`, so my goal was to contribute to `py-air-control` and `coapthon3`, and make sure tests are included in the PyPi distribution.

Fetching from GitHub is easier, so I'll do that first. I can always migrate back to `fetchPypi` once the tests are distributed.

I just hope their tests can run in a sandbox. :crossed_fingers:

Comment (urbas) [reply pkgs/development/python-modules/py-air-control/default.nix]: I now added tests, but there were a few complications:
1. A couple of tests required special devices and one required port redirection (I disabled and commented on them).
2. Tests take a very long time. I added a `pytest` option to print out how long each individual test takes. I'm at an impasse here and would like to ask for your advice on this. Should I disable some tests? If so, which ones? Could you please take a look at the build output and look at the durations?

**Comment by SuperSandro2000** [comment]
On darwin it gets stuck at 

```
[1/0/5 built, 0.0 MiB DL] building python3.7-CoAPthon3-1.0.1 (pytestCheckPhase): ======================= 38 failed, 2 deselected in 2.36s =======================
```

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 105078` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python37Packages.coapthon3</li>
    <li>python37Packages.py-air-control</li>
    <li>python38Packages.coapthon3</li>
    <li>python38Packages.py-air-control</li>
  </ul>
</details>

**Comment by urbas** [comment]
> On darwin it gets stuck at
> 
> ```
> [1/0/5 built, 0.0 MiB DL] building python3.7-CoAPthon3-1.0.1 (pytestCheckPhase): ======================= 38 failed, 2 deselected in 2.36s =======================
> ```

I have seen this happen on x86 before I disabled the two failing tests. I think there's something wrong with the tear-down logic when tests fail. Note that tests create some servers, which create a bunch of threads. So maybe the failed tests leave behind a lingering [non-daemon thread](https://github.com/Tanganelli/CoAPthon3/blob/89d51737a9f95a0dbd113a27f4ba88fe6230859f/coapthon/server/coap.py#L50-L51) that prevents pytest to shut down properly.

In any case, since all these tests fail on darwin and they take more than 10 minutes I will just disable them.

I will try to contribute some fast and robust unit tests to both `coapthon3` and `py-air-control` projects. We can then invoke only the unit tests rather than these slow and brittle integration tests.

**Comment by urbas** [comment]
Btw, I now also verified this builds on aarch64 (NixOS on Raspberry Pi 4).

**Review by lukegb — APPROVED** [review]

**Comment by lukegb** [comment]
Result of `nixpkgs-review pr 105078` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python37Packages.coapthon3</li>
    <li>python37Packages.py-air-control</li>
    <li>python38Packages.coapthon3</li>
    <li>python38Packages.py-air-control</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
> Tests create some servers, which in turn create a bunch of threads. Maybe the failed tests leave behind some lingering non-daemon threads which prevent pytest to shut down properly.
>
> In any case, since all these tests fail on darwin and they take more than 10 minutes I will just disable them.

Yeah, we do not want to waste reviewers and builders time with that. 

> I will try to contribute some fast and robust unit tests to both `coapthon3` and `py-air-control` projects. We can then invoke only the unit tests rather than these slow and brittle integration tests.

This won't block the PR but great that you try to improve the project.

**Comment by SuperSandro2000** [comment]
@urbas I think we need to do the same with the other package.

```
[1/1/5 built, 2 copied (71.0/71.0 MiB), 19.3 MiB DL] building python3.7-py-air-control-2.1.0 (pytestCheckPhase): ====================== 12 deselected, 20 errors in 1.84s ==================
```

**Comment by urbas** [comment]
> @urbas I think we need to do the same with the other package.
> 
> ```
> [1/1/5 built, 2 copied (71.0/71.0 MiB), 19.3 MiB DL] building python3.7-py-air-control-2.1.0 (pytestCheckPhase): ====================== 12 deselected, 20 errors in 1.84s ==================
> ```

Oh smokes, I hoped at least those would pass. I disabled them now. :+1:

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 105078` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>python37Packages.coapthon3</li>
    <li>python37Packages.py-air-control</li>
    <li>python38Packages.coapthon3</li>
    <li>python38Packages.py-air-control</li>
    <li>python39Packages.coapthon3</li>
    <li>python39Packages.py-air-control</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 105078` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>python37Packages.coapthon3</li>
    <li>python37Packages.py-air-control</li>
    <li>python38Packages.coapthon3</li>
    <li>python38Packages.py-air-control</li>
    <li>python39Packages.coapthon3</li>
    <li>python39Packages.py-air-control</li>
  </ul>
</details>

**Pull request merged**